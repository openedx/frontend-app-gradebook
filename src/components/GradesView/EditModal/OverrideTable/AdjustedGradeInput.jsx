/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Form } from '@edx/paragon';

import selectors from 'data/selectors';
import actions from 'data/actions';
import { getLocale, isRtl } from '@edx/frontend-platform/i18n';

/**
 * <AdjustedGradeInput />
 * Input control for adjusting the grade of a unit
 * displays an "/ ${possibleGrade} if there is one in the data model.
 */
export class AdjustedGradeInput extends React.Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
  }

  onChange = ({ target }) => {
    let adjustedGradeValue;
    switch (true) {
      case target.value < 0:
        adjustedGradeValue = 0;
        break;
      case this.props.possibleGrade && target.value > this.props.possibleGrade:
        adjustedGradeValue = this.props.possibleGrade;
        break;
      default:
        adjustedGradeValue = target.value;
    }
    this.props.setModalState({ adjustedGradeValue });
  };

  render() {
    return (
      <span>
        <Form.Control
          type="number"
          name="adjustedGradeValue"
          min="0"
          max={this.props.possibleGrade ? this.props.possibleGrade : ''}
          value={this.props.value}
          onChange={this.onChange}
        />
        {this.props.possibleGrade && ` ${isRtl(getLocale()) ? '\\' : '/'} ${this.props.possibleGrade}`}
      </span>
    );
  }
}
AdjustedGradeInput.defaultProps = {
  possibleGrade: null,
};
AdjustedGradeInput.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  possibleGrade: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  setModalState: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  possibleGrade: selectors.root.editModalPossibleGrade(state),
  value: selectors.app.modalState.adjustedGradeValue(state),
});

export const mapDispatchToProps = {
  setModalState: actions.app.setModalState,
};

export default connect(mapStateToProps, mapDispatchToProps)(AdjustedGradeInput);
