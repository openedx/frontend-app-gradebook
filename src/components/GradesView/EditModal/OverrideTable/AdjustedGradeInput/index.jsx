import React from 'react';

import { Form } from '@edx/paragon';

import useAdjustedGradeInputData from './hooks';

/**
 * <AdjustedGradeInput />
 * Input control for adjusting the grade of a unit
 * displays an "/ ${possibleGrade} if there is one in the data model.
 */
export const AdjustedGradeInput = () => {
  const {
    value,
    onChange,
    hintText,
  } = useAdjustedGradeInputData();
  return (
    <span>
      <Form.Control
        type="text"
        name="adjustedGradeValue"
        value={value}
        onChange={onChange}
      />
      {hintText}
    </span>
  );
};

AdjustedGradeInput.propTypes = {};

export default AdjustedGradeInput;
