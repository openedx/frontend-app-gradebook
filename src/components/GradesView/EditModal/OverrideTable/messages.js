import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  adjustedGradeHeader: {
    id: 'gradebook.GradesView.EditModal.Overrides.adjustedGradeHeader',
    defaultMessage: 'Adjusted grade',
    description: 'Edit Modal Override Table Adjusted grade column header',
  },
  dateHeader: {
    id: 'gradebook.GradesView.EditModal.Overrides.dateHeader',
    defaultMessage: 'Date',
    description: 'Edit Modal Override Table Date column header',
  },
  graderHeader: {
    id: 'gradebook.GradesView.EditModal.Overrides.graderHeader',
    defaultMessage: 'Grader',
    description: 'Edit Modal Override Table Grader column header',
  },
  reasonHeader: {
    id: 'gradebook.GradesView.EditModal.Overrides.reasonHeader',
    defaultMessage: 'Reason',
    description: 'Edit Modal Override Table Reason column header',
  },
});

export default messages;
