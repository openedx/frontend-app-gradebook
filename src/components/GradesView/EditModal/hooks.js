import { selectors, actions, thunkActions } from 'data/redux/hooks';

export const useEditModalData = () => {
  const error = selectors.grades.useGradeData().gradeOverrideHistoryError;
  const isOpen = selectors.app.useModalData().open;
  const closeModal = actions.app.useCloseModal();
  const doneViewingAssignment = actions.grades.useDoneViewingAssignment();
  const updateGrades = thunkActions.grades.useUpdateGrades();

  const onClose = () => {
    doneViewingAssignment();
    closeModal();
  };

  const handleAdjustedGradeClick = () => {
    updateGrades();
    doneViewingAssignment();
    closeModal();
  };

  return {
    onClose,
    error,
    handleAdjustedGradeClick,
    isOpen,
  };
};

export default useEditModalData;
