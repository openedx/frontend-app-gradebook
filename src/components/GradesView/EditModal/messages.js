import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  assignmentHeader: {
    id: 'gradebook.GradesView.EditModal.headers.assignment',
    defaultMessage: 'Assignment',
    description: 'Edit Modal Assignment header',
  },
  currentGradeHeader: {
    id: 'gradebook.GradesView.EditModal.headers.currentGrade',
    defaultMessage: 'Current Grade',
    description: 'Edit Modal Current Grade header',
  },
  originalGradeHeader: {
    id: 'gradebook.GradesView.EditModal.headers.originalGrade',
    defaultMessage: 'Original Grade',
    description: 'Edit Modal Original Grade header',
  },
  studentHeader: {
    id: 'gradebook.GradesView.EditModal.headers.student',
    defaultMessage: 'Student',
    description: 'Edit Modal Student header',
  },
  title: {
    id: 'gradebook.GradesView.EditModal.title',
    defaultMessage: 'Edit Grades',
    description: 'Edit Modal title',
  },
  closeText: {
    id: 'gradebook.GradesView.EditModal.closeText',
    defaultMessage: 'Cancel',
    description: 'Edit Modal close button text',
  },
  visibility: {
    id: 'gradebook.GradesView.EditModal.contactSupport',
    defaultMessage: 'Showing most recent actions (max 5). To see more, please contact support',
    description: 'Edit Modal visibility hint message',
  },
  saveVisibility: {
    id: 'gradebook.GradesView.EditModal.saveVisibility',
    defaultMessage: 'Note: Once you save, your changes will be visible to students.',
    description: 'Edit Modal saved changes effect hint',
  },
  saveGrade: {
    id: 'gradebook.GradesView.EditModal.saveGrade',
    defaultMessage: 'Save Grades',
    description: 'Edit Modal Save button label',
  },
});

export default messages;
