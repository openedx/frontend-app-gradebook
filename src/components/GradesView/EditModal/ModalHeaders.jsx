import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { FormattedMessage } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';

import messages from './messages';
import HistoryHeader from './HistoryHeader';

/**
 * <ModalHeaders />
 * Provides a list of HistoryHeaders for the student name, assignment,
 * original grade, and current override grade.
 */
export const ModalHeaders = ({
  modalState,
  originalGrade,
  currentGrade,
}) => (
  <div>
    <HistoryHeader
      id="assignment"
      label={<FormattedMessage {...messages.assignmentHeader} />}
      value={modalState.assignmentName}
    />
    <HistoryHeader
      id="student"
      label={<FormattedMessage {...messages.studentHeader} />}
      value={modalState.updateUserName}
    />
    <HistoryHeader
      id="original-grade"
      label={<FormattedMessage {...messages.originalGradeHeader} />}
      value={originalGrade}
    />
    <HistoryHeader
      id="current-grade"
      label={<FormattedMessage {...messages.currentGradeHeader} />}
      value={currentGrade}
    />
  </div>
);
ModalHeaders.defaultProps = {
  currentGrade: null,
  originalGrade: null,
};
ModalHeaders.propTypes = {
  // redux
  currentGrade: PropTypes.number,
  originalGrade: PropTypes.number,
  modalState: PropTypes.shape({
    assignmentName: PropTypes.string.isRequired,
    updateUserName: PropTypes.string,
  }).isRequired,
};

export const mapStateToProps = (state) => ({
  modalState: {
    assignmentName: selectors.app.modalState.assignmentName(state),
    updateUserName: selectors.app.modalState.updateUserName(state),
  },
  currentGrade: selectors.grades.gradeOverrideCurrentEarnedGradedOverride(state),
  originalGrade: selectors.grades.gradeOriginalEarnedGraded(state),
});

export default connect(mapStateToProps)(ModalHeaders);
