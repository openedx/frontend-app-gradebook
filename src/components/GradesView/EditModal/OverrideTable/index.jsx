/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { DataTable } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import { gradeOverrideHistoryColumns as columns } from 'data/constants/app';
import selectors from 'data/selectors';

import messages from './messages';
import ReasonInput from './ReasonInput';
import AdjustedGradeInput from './AdjustedGradeInput';

/**
 * <OverrideTable />
 * Table containing previous grade override entries, and an "edit" row
 * with todays date, an AdjustedGradeInput and a ReasonInput
 */
export const OverrideTable = ({
  hide,
  gradeOverrides,
  todaysDate,
}) => {
  if (hide) {
    return null;
  }
  return (
    <DataTable
      columns={[
        { Header: <FormattedMessage {...messages.dateHeader} />, accessor: columns.date },
        { Header: <FormattedMessage {...messages.graderHeader} />, accessor: columns.grader },
        { Header: <FormattedMessage {...messages.reasonHeader} />, accessor: columns.reason },
        {
          Header: <FormattedMessage {...messages.adjustedGradeHeader} />,
          accessor: columns.adjustedGrade,
        },
      ]}
      data={[
        ...gradeOverrides,
        {
          adjustedGrade: <AdjustedGradeInput />,
          date: todaysDate,
          reason: <ReasonInput />,
        },
      ]}
      itemCount={gradeOverrides.length}
    />
  );
};
OverrideTable.defaultProps = {
  gradeOverrides: [],
};
OverrideTable.propTypes = {
  // redux
  gradeOverrides: PropTypes.arrayOf(PropTypes.shape({
    date: PropTypes.string,
    grader: PropTypes.string,
    reason: PropTypes.string,
    adjustedGrade: PropTypes.number,
  })),
  hide: PropTypes.bool.isRequired,
  todaysDate: PropTypes.string.isRequired,
};

export const mapStateToProps = (state) => ({
  hide: selectors.grades.hasOverrideErrors(state),
  gradeOverrides: selectors.grades.gradeOverrides(state),
  todaysDate: selectors.app.modalState.todaysDate(state),
});

export default connect(mapStateToProps)(OverrideTable);
