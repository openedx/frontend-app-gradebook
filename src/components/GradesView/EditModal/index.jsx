/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  Button,
  Modal,
  Alert,
} from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';
import actions from 'data/actions';
import thunkActions from 'data/thunkActions';

import messages from './messages';
import OverrideTable from './OverrideTable';
import ModalHeaders from './ModalHeaders';

/**
 * <EditModal />
 * Wrapper component for the modal that allows editing the grade for an individual
 * unit, for a given student.
 * Provides a StatusAlert with override fetch errors if any are found, an OverrideTable
 * (with appropriate headers) for managing the actual override, and a submit button for
 * adjusting the grade.
 * (also provides a close button that clears the modal state)
 */
export class EditModal extends React.Component {
  constructor(props) {
    super(props);
    this.closeAssignmentModal = this.closeAssignmentModal.bind(this);
    this.handleAdjustedGradeClick = this.handleAdjustedGradeClick.bind(this);
  }

  closeAssignmentModal() {
    this.props.doneViewingAssignment();
    this.props.closeModal();
  }

  handleAdjustedGradeClick() {
    this.props.updateGrades();
    this.closeAssignmentModal();
  }

  render() {
    return (
      <Modal
        open={this.props.open}
        title={<FormattedMessage {...messages.title} />}
        closeText={<FormattedMessage {...messages.closeText} />}
        body={(
          <div>
            <ModalHeaders />
            <Alert
              variant="danger"
              show={!!this.props.gradeOverrideHistoryError}
              dismissible={false}
            >
              {this.props.gradeOverrideHistoryError}
            </Alert>
            <OverrideTable />
            <div><FormattedMessage {...messages.visibility} /></div>
            <div><FormattedMessage {...messages.saveVisibility} /></div>
          </div>
        )}
        buttons={[
          <Button variant="primary" onClick={this.handleAdjustedGradeClick}>
            <FormattedMessage {...messages.saveGrade} />
          </Button>,
        ]}
        onClose={this.closeAssignmentModal}
      />
    );
  }
}

EditModal.defaultProps = {
  gradeOverrideHistoryError: '',
};

EditModal.propTypes = {
  // redux
  gradeOverrideHistoryError: PropTypes.string,
  open: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  doneViewingAssignment: PropTypes.func.isRequired,
  updateGrades: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  gradeOverrideHistoryError: selectors.grades.gradeOverrideHistoryError(state),
  open: selectors.app.modalState.open(state),
});

export const mapDispatchToProps = {
  closeModal: actions.app.closeModal,
  doneViewingAssignment: actions.grades.doneViewingAssignment,
  updateGrades: thunkActions.grades.updateGrades,
};

export default connect(mapStateToProps, mapDispatchToProps)(EditModal);
