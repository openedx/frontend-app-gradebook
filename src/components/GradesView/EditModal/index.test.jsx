import React from 'react';
import { shallow } from 'enzyme';

import {
  ActionRow,
  ModalDialog,
} from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { formatMessage } from 'testUtils';

import ModalHeaders from './ModalHeaders';
import OverrideTable from './OverrideTable';
import useEditModalData from './hooks';
import EditModal from '.';
import messages from './messages';

jest.mock('./hooks', () => jest.fn());
jest.mock('./ModalHeaders', () => 'ModalHeaders');
jest.mock('./OverrideTable', () => 'OverrideTable');

const hookProps = {
  onClose: jest.fn().mockName('hooks.onClose'),
  error: 'test-error',
  handleAdjustedGradeClick: jest.fn().mockName('hooks.handleAdjustedGradeClick'),
  isOpen: 'test-is-open',
};
useEditModalData.mockReturnValue(hookProps);

let el;
describe('EditModal component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    el = shallow(<EditModal />);
  });
  describe('behavior', () => {
    it('initializes intl hook', () => {
      expect(useIntl).toHaveBeenCalled();
    });
    it('initializes component hooks', () => {
      expect(useEditModalData).toHaveBeenCalled();
    });
  });
  describe('render', () => {
    test('modal props', () => {
      const modalProps = el.find(ModalDialog).props();
      expect(modalProps.title).toEqual(formatMessage(messages.title));
      expect(modalProps.isOpen).toEqual(hookProps.isOpen);
      expect(modalProps.onClose).toEqual(hookProps.onClose);
    });
    const loadBody = () => {
      const body = el.find(ModalDialog).children().at(0);
      const children = body.find('div').children();
      return { body, children };
    };
    const testBody = () => {
      test('type', () => {
        const { body } = loadBody();
        expect(body.type()).toEqual('ModalDialog.Body');
      });
      test('headers row', () => {
        const { children } = loadBody();
        expect(children.at(0)).toMatchObject(shallow(<ModalHeaders />));
      });
      test('table row', () => {
        const { children } = loadBody();
        expect(children.at(2)).toMatchObject(shallow(<OverrideTable />));
      });
      test('messages', () => {
        const { children } = loadBody();
        expect(
          children.at(3).contains(formatMessage(messages.visibility)),
        ).toEqual(true);
        expect(
          children.at(4).contains(formatMessage(messages.saveVisibility)),
        ).toEqual(true);
      });
    };
    const testFooter = () => {
      let footer;
      beforeEach(() => {
        footer = el.find(ModalDialog).children().at(1);
      });
      test('type', () => {
        expect(footer.type()).toEqual('ModalDialog.Footer');
      });
      test('contains action row', () => {
        expect(footer.children().at(0).type()).toEqual('ActionRow');
      });
      test('close button', () => {
        const button = footer.find(ActionRow).children().at(0);
        expect(button.contains(formatMessage(messages.closeText))).toEqual(true);
        expect(button.type()).toEqual('ModalDialog.CloseButton');
      });
      test('adjusted grade button', () => {
        const button = footer.find(ActionRow).children().at(1);
        expect(button.contains(formatMessage(messages.saveGrade))).toEqual(true);
        expect(button.type()).toEqual('Button');
        expect(button.props().onClick).toEqual(hookProps.handleAdjustedGradeClick);
      });
    };
    describe('without error', () => {
      beforeEach(() => {
        useEditModalData.mockReturnValueOnce({ ...hookProps, error: undefined });
        el = shallow(<EditModal />);
      });
      test('snapshot', () => {
        expect(el).toMatchSnapshot();
      });
      testBody();
      testFooter();
      test('alert row', () => {
        const alert = loadBody().children.at(1);
        expect(alert.type()).toEqual('Alert');
        expect(alert.props().show).toEqual(false);
      });
    });
    describe('with error', () => {
      test('snapshot', () => {
        expect(el).toMatchSnapshot();
      });
      testBody();
      test('alert row', () => {
        const alert = loadBody().children.at(1);
        expect(alert.type()).toEqual('Alert');
        expect(alert.props().show).toEqual(true);
        expect(alert.contains(hookProps.error)).toEqual(true);
      });
      testFooter();
    });
  });
});
