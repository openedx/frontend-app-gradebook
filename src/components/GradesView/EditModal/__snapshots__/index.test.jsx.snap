// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditModal component render with error snapshot 1`] = `
<ModalDialog
  hasCloseButton={true}
  isFullscreenOnMobile={true}
  isOpen="test-is-open"
  onClose={[MockFunction hooks.onClose]}
  size="xl"
  title="Edit Grades"
>
  <ModalDialog.Body>
    <div>
      <ModalHeaders />
      <Alert
        dismissible={false}
        show={true}
        variant="danger"
      >
        test-error
      </Alert>
      <OverrideTable />
      <div>
        Showing most recent actions (max 5). To see more, please contact support
      </div>
      <div>
        Note: Once you save, your changes will be visible to students.
      </div>
    </div>
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        variant="tertiary"
      >
        Cancel
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction hooks.handleAdjustedGradeClick]}
        variant="primary"
      >
        Save Grades
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;

exports[`EditModal component render without error snapshot 1`] = `
<ModalDialog
  hasCloseButton={true}
  isFullscreenOnMobile={true}
  isOpen="test-is-open"
  onClose={[MockFunction hooks.onClose]}
  size="xl"
  title="Edit Grades"
>
  <ModalDialog.Body>
    <div>
      <ModalHeaders />
      <Alert
        dismissible={false}
        show={false}
        variant="danger"
      />
      <OverrideTable />
      <div>
        Showing most recent actions (max 5). To see more, please contact support
      </div>
      <div>
        Note: Once you save, your changes will be visible to students.
      </div>
    </div>
  </ModalDialog.Body>
  <ModalDialog.Footer>
    <ActionRow>
      <ModalDialog.CloseButton
        variant="tertiary"
      >
        Cancel
      </ModalDialog.CloseButton>
      <Button
        onClick={[MockFunction hooks.handleAdjustedGradeClick]}
        variant="primary"
      >
        Save Grades
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</ModalDialog>
`;
